include/nighthawk/common/factories.h

```diff
@@ -33,12 +33,21 @@ class StatisticFactory {
   virtual StatisticPtr create() const PURE;
 };
 
+class RequestSourceConstructorInterface {
+public:
+  virtual ~RequestSourceConstructorInterface() = default;
+  virtual RequestSourcePtr
+  createStaticRequestSource(Envoy::Http::RequestHeaderMapPtr&&,
+                            const uint64_t max_yields = UINT64_MAX) const PURE;
```

Comment:
Somehow specifying the default value in the interface feels off to me / implementation-like. But then this interface seems to target materializing concrete implementations of `RequestSourceContructor`. I'm not sure I fully understand the rationale here, could you help me there by sharing some context? My concern is that for each implementation we will have to add to this interface, which somehow feel counterintuitive.

Reply:
This is part of a wider set of changes I'm making for the request source plugin implementation. So ideally, once everything is in plugin form, the RequestSourceConstructor interface will be fairly stable. The reason I am splitting this off is to have better separation of logic between the decision of what kind of requestsource I am making, and how to make that request source. This also has the added benefit of making things more testable.

Reply:
I wonder: What if the factories 'create' methode would be modified instead to take a proto config as an argument? And then its imementation instantiate  the right class  based that? (The proto config would use a envoy.config.core.v3.TypedExtensionConfig to indicate the desired request source)

Reply:
I think I agree with Otto here.  Rather than having two different functions for createRemoteRequestSource and createStaticRequestSource, just have the interface contain one `create` function, that includes a proto for specific configuration, based on what people want to create.

Right now, the difference between these two methods appears to be the inclusion of max_yields.  I might not really understand what max_yields is doing here, but it seems likely that would be an implementation detail of specific implementations of this interface, rather than a part of the interface itself.  Otto's proto solution handles this by letting those implementations define max_yields or anything else they might not need.